File: coolrmi/CoolRMI.Net/Messages/CoolRMICreateProxy.cs
using System;
using System.IO;
using CoolRMI.Net.Serializer;

namespace CoolRMI.Net.Messages
{
    [Serializable]
    public class CoolRMICreateProxy : AbstractCoolRMIMessage, ICustomPortableSerializable
    {
        public long ProxyId { get; }
        public Type Iface { get; }
        public override string Name => ToString();

        public CoolRMICreateProxy(long queryId, long proxyId,
            Type iface) : base(queryId)
        {
            ProxyId = proxyId;
            Iface = iface;
        }

        public override string ToString()
        {
            return "Create proxy: " + ProxyId + " " + Iface.AssemblyQualifiedName;
        }

        // ReSharper disable once UnusedMember.Local
        private CoolRMICreateProxy(PortableSerializer serializer,
            BinaryReader br)
        {
            ObjectSerializer.DefaultDeserialize(serializer, br, this,
                typeof(AbstractCoolRMIMessage));
            ProxyId = br.ReadInt64();
            Iface = serializer.ReadClassName(br);
        }

        void ICustomPortableSerializable.Serialize(
            PortableSerializer serializer, BinaryWriter bw)
        {
            ObjectSerializer.DefaultSerialize(serializer, bw, this,
                typeof(AbstractCoolRMIMessage));
            bw.Write(ProxyId);
            serializer.WriteClassName(bw, Iface);
        }
    }

    [Serializable]
    public class CoolRMICreateProxyReply : AbstractCoolRMIReply
    {
        public override string Name => ToString();

        public CoolRMICreateProxyReply(long queryId) : base(queryId) {}

        public override string ToString()
        {
            return "proxy created";
        }
    }
}

File: coolrmi/hu.qgears.coolrmi/src/hu/qgears/coolrmi/messages/CoolRMICreateProxy.java
package hu.qgears.coolrmi.messages;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.Serializable;

import hu.qgears.coolrmi.serializer.ICustomPortableSerializable;
import hu.qgears.coolrmi.serializer.JavaType;
import hu.qgears.coolrmi.serializer.ObjectSerializer;
import hu.qgears.coolrmi.serializer.PortableSerializer;
import hu.qgears.coolrmi.serializer.Utils;


public class CoolRMICreateProxy
	extends AbstractCoolRMIMessage
	implements Serializable, ICustomPortableSerializable {
	private static final long serialVersionUID = 1L;
	private long proxyId;
	private String ifaceName;
	public String getIfaceName() {
		return ifaceName;
	}

	public long getProxyId() {
		return proxyId;
	}

	public CoolRMICreateProxy() {
		
	}

	public CoolRMICreateProxy(long queryId, 
			long proxyId,
			String ifaceName) {
		super(queryId);
		this.proxyId = proxyId;
		this.ifaceName=ifaceName;
	}

	@Override
	public String toString() {
		return "create proxy "+proxyId+" "+ifaceName;
	}
	@Override
	public String getName() {
		return "Create proxy: "+proxyId+" "+ifaceName;
	}

	@SuppressWarnings("unused") // reflection
	private CoolRMICreateProxy(PortableSerializer serializer, InputStream is)
			throws Exception {
		ObjectSerializer.defaultDeserialize(serializer, is, this,
				AbstractCoolRMIMessage.class);
		proxyId = Utils.read64(is);
		ifaceName = serializer.readClassName(is).getCls().getName();
	}

	@Override
	public void serialize(PortableSerializer serializer, OutputStream os)
			throws IOException {
		ObjectSerializer.defaultSerialize(serializer, os, this,
				AbstractCoolRMIMessage.class);
		Utils.write64(os, proxyId);
		try {
			serializer.writeClassName(os,
					new JavaType(Class.forName(ifaceName)));
		} catch (ClassNotFoundException e) {
			throw new RuntimeException(e);
		}
	}
}
